parlai/agents/memnn/memnn.py

```diff
@@ -119,10 +115,13 @@ def observe(self, observation):
             # if the last example wasn't the end of an episode, then we need to
             # recall what was said in that example
             prev_dialogue = self.observation['text'] if self.observation is not None else ''
-            batch_idx = self.opt.get('batchindex', 0)
-            if self.answers[batch_idx] is not None:
-                prev_dialogue += '\n' + self.answers[batch_idx]
-                self.answers[batch_idx] = None
+
+            # append answer given in the previous example to the dialog 
+            if 'eval_labels' in self.observation:
+                prev_dialogue += '\n' + random.choice(self.observation['eval_labels'])
+            elif 'labels' in self.observation:
+                prev_dialogue += '\n' + random.choice(self.observation['labels'])
+
```

Comment:
@jaseweston in the seq2seq model we were doing 'append label during training but append answer during valid' but maybe this makes more sense--append labels if available, else append eval_labels if available, else append self answer, even? (this will make it more robust during interactive modes as well)

Reply:
we should be using my utility. yes, we could have an option where it uses label in train and pred when no label or eval_label: these should be added to the utility as extra options i think.

Reply:
do you want to land this or wait for a switch to your util?

Reply:
You're talking about maintain_dialog_history, right? It seems like it can use as reply either self.answer (predictions from the previous example) or the current label (the one we're trying to predict) or none. Am I missing something? Shouldn't it use labels from the previous example?

Reply:
Hmm, that does look wrong.. needs to be fixed I think

Reply:
should be like this i think: https://github.com/facebookresearch/ParlAI/pull/461

?

Reply:
Maybe also add a check if eval_labels become labels and vice versa? to fix that training data is appended to validation data when a model is evaluated in the middle of training

Reply:
not sure that is the job of this function, but that should be definitely fixed -- are you aware of that problem @alexholdenmiller  et al?